libethereum/Precompiled.cpp

```diff
@@ -44,47 +47,48 @@ static bytes ecrecoverCode(bytesConstRef _in)
 	memcpy(&in, _in.data(), min(_in.size(), sizeof(in)));
 
 	h256 ret;
-
-	if ((u256)in.v > 28)
-		return ret.asBytes();
-	SignatureStruct sig(in.r, in.s, (byte)((int)(u256)in.v - 27));
-	if (!sig.isValid())
-		return ret.asBytes();
-
-	try
+	if ((u256)in.v <= 28)
 	{
-		ret = dev::sha3(recover(sig, in.hash));
+		SignatureStruct sig(in.r, in.s, (byte)((int)(u256)in.v - 27));
+		if (sig.isValid())
+		{
+			try
+			{
+				ret = sha3(recover(sig, in.hash));
+			}
+			catch (...) {}
+		}
 	}
-	catch (...) {}
 
 	memset(ret.data(), 0, 12);
-	return ret.asBytes();
+	ret.ref().copyTo(_out);
 }
 
-static bytes sha256Code(bytesConstRef _in)
+void sha256Code(bytesConstRef _in, bytesRef _out)
 {
-	return sha256(_in).asBytes();
+	sha256(_in).ref().copyTo(_out);
 }
 
-static bytes ripemd160Code(bytesConstRef _in)
+void ripemd160Code(bytesConstRef _in, bytesRef _out)
 {
-	return h256(ripemd160(_in), h256::AlignRight).asBytes();
+	h256(ripemd160(_in), h256::AlignRight).ref().copyTo(_out);
 }
 
-static bytes identityCode(bytesConstRef _in)
+void identityCode(bytesConstRef _in, bytesRef _out)
 {
-	return _in.toBytes();
+	_in.copyTo(_out);
 }
 
-static const std::unordered_map<unsigned, PrecompiledAddress> c_precompiled =
-{
-	{ 1, { [](bytesConstRef) -> bigint { return c_ecrecoverGas; }, ecrecoverCode }},
-	{ 2, { [](bytesConstRef i) -> bigint { return c_sha256Gas + (i.size() + 31) / 32 * c_sha256WordGas; }, sha256Code }},
-	{ 3, { [](bytesConstRef i) -> bigint { return c_ripemd160Gas + (i.size() + 31) / 32 * c_ripemd160WordGas; }, ripemd160Code }},
-	{ 4, { [](bytesConstRef i) -> bigint { return c_identityGas + (i.size() + 31) / 32 * c_identityWordGas; }, identityCode }}
-};
+}
 
 std::unordered_map<unsigned, PrecompiledAddress> const& dev::eth::precompiled()
 {
-	return c_precompiled;
+	static const std::unordered_map<unsigned, PrecompiledAddress> s_precompiled =
```

Comment:
Ok. I thought `c_` is for compile-time constants only.

I moved it inside the function because C++11 guarantees thread-safety in this case.